frontend/src/metabase-lib/expressions/diagnostics.ts

```diff
@@ -88,19 +90,61 @@ export function diagnose({
       : mismatchedParentheses < -1
       ? t`Expecting ${-mismatchedParentheses} opening parentheses`
       : null;
+
   if (message) {
     return { message };
   }
 
+  // try to compile on FE
+  let mbqlOrError: Expr | ErrorWithMessage;
   try {
-    return prattCompiler({ source, startRule, name, query, stageIndex });
+    mbqlOrError = prattCompiler({ source, startRule, name, query, stageIndex });
+
+    if (isErrorWithMessage(mbqlOrError)) {
+      return mbqlOrError;
+    }
+
+    if (startRule === "expression" && isBooleanExpression(mbqlOrError)) {
+      throw new ResolverError(
+        t`Custom columns do not support boolean expressions`,
+        mbqlOrError.node,
+      );
+    }
   } catch (err) {
+    console.warn("compile error", err);
+
     if (isErrorWithMessage(err)) {
       return err;
     }
 
     return { message: t`Invalid expression` };
   }
+
+  // now diagnose on BE
+  const startRuleToExpressionModeMapping: Record<string, Lib.ExpressionMode> = {
+    boolean: "filter",
+  };
+
+  const expressionMode: Lib.ExpressionMode =
+    startRuleToExpressionModeMapping[startRule] ?? startRule;
+
+  const possibleError = Lib.diagnoseExpression(
+    query,
+    stageIndex,
+    expressionMode,
+    mbqlOrError,
+    expressionPosition,
+  );
+
+  if (possibleError) {
+    console.warn("diagnostic error", possibleError.message);
```

Comment:
`possibleError.message` is a translated message, so I wonder if we can do something better than just throwing away.

Reply:
What about keeping the part before the colon? I think `Type error` might help a bit. And it's very different from the error you get if you try to create circular definitions.

Reply:
I don't think `Type error` is good enough for the user